R/utilities-units.R

```diff
@@ -372,26 +372,105 @@ getDimensionByName <- function(name) {
   enum(allAvailableDimensions())
 }
 
+
 #' @title Supported dimensions defined as a named list
 #'
 #' @details
 #' ospDimensions$Mass => "Mass"
 #'
 #' @export
-ospDimensions <- list()
+ospDimensions <- NULL
 
 #' Supported units defined as a named list of lists
 #'
 #' ospUnits$Mass$kg => "kg"
 #' @export
-ospUnits <- list()
+ospUnits <- NULL
+
+#' parse OSPSuite.Dimensions.xml containing dimensions and units
+#'
+#' @return An XML document
+#' @import xml2
+#'
+#' @examples
+.parseDimensionsXML <- function() {
+  # Read the XML file
+  xmlFile <- system.file("lib/OSPSuite.Dimensions.xml", package = "ospsuite")
+  xmlData <- xml2::read_xml(xmlFile)
+
+  return(xmlData)
+}
+
+#' get OSP Dimensions from OSPSuite.Dimensions.xml data
+#'
+#' @param xmlData XML data from `.parseDimensionsXML()`
+#'
+#' @return a list of supported dimensions
+.getOspDimensions <- function(xmlData) {
+  ospDimensions <- list()
+
+  dimensionsNodes <- xmlData %>%
+    xml2::xml_find_all(".//Dimension")
+
+  for (dimNode in dimensionsNodes) {
+    name <- xml2::xml_attr(dimNode, "name")
+    ospDimensions[[name]] <- name
+  }
+
+  ospDimensions[["Dimensionless"]] <- "Dimensionless"
+
+  ospDimensions <- ospDimensions[order(names(ospDimensions))]
+
+  return(ospDimensions)
+}
+
+#' get OSP Units from OSPSuite.Dimensions.xml data
+#'
+#' @param xmlData XML data from `.parseDimensionsXML()`
+#'
+#' @return a list of supported units
+.getOspUnits <- function(xmlData) {
+  # Extract information from the XML
+  dimensionsNodes <- xmlData %>%
+    xml2::xml_find_all(".//Dimension")
+
+  ospUnits <- list()
+
+  for (dim in dimensionsNodes) {
+    dim_name <- dim %>%
+      xml2::xml_attr("name") %>%
+      gsub(pattern = "[(]", replacement = "[") %>%
+      gsub(pattern = "[)]", replacement = "]")
+    dim_units <- dim %>%
+      xml2::xml_find_all(".//Unit")
+
+    unit_list <- list()
+
+    for (unit in dim_units) {
+      unit_name <- unit %>% xml2::xml_attr("name")
+      # if unit_name equals "" replace by Unitless
+      if (unit_name == "") {
+        unit_name <- "Unitless"
+      }
+      unit_list[[unit_name]] <- unit_name
+    }
+    ospUnits[[dim_name]] <- unit_list
+  }
+
+  ospUnits[["Dimensionless"]] <- list("Unitless" = "Unitless")
```

Comment:
why do we actually replace the empty unit with this funny "Unitless"?
it seems to be unused at the moment (at least I could not find any usage of "Unitless" within the package), but if we would pass e.g. `ospUnits$Dimensionless$Unitless` to OSPSuiteCore it would crash(??)

Reply:
currently:
```
> ospUnits$Dimensionless$Unitless
[1] "Unitless"
> ospUnits$Fraction$Unitless
[1] "Unitless"
```

my expectation:
```
> ospUnits$Dimensionless$Unitless
[1] ""
> ospUnits$Fraction$Unitless
[1] ""
```

Reply:
I really only replicated the output of the previous `.getUnitsEnum()`.

But my guess would be that for plots, displaying "" could be misunderstood for a bug when the user needs to show units.

Reply:
ok, created a new issue for that: https://github.com/Open-Systems-Pharmacology/OSPSuite-R/issues/1328